### PR TITLE
docs: document semaphores

### PR DESCRIPTION
**Pull request by bhumitattarde**

### CODE REVIEW

File: libs/core/synchronization/include/hpx/synchronization/counting_semaphore.hpp [hunk 1]
```diff
@@ -22,41 +22,379 @@
...
+        ///         \a try_wait returns false if not sufficient credits are
+        ///         available at this point in time.
+        ///
+        bool try_wait(std::ptrdiff_t count = 1);
+
+        ///
+        /// \brief Signal the semaphore.
+        ///
+        /// \param count The value by which the internal lock count will be
+        ///              incremented.
```
Comment (bhumitattarde) [inline libs/core/synchronization/include/hpx/synchronization/counting_semaphore.hpp:364]: @hkaiser is this correct?

File: libs/core/synchronization/include/hpx/synchronization/counting_semaphore.hpp [hunk 2]
```diff
@@ -22,41 +22,379 @@
...
+        ///
+        void signal(std::ptrdiff_t count = 1);
+
+        ///
+        /// \brief Unblock all acquirers.
+        ///
+        /// \return std::ptrdiff_t internal lock count after the operation.
+        ///
+        std::ptrdiff_t signal_all();
```
Comment (bhumitattarde) [inline libs/core/synchronization/include/hpx/synchronization/counting_semaphore.hpp:373]: @hkaiser is this documentation correct or would you word it differently? Also, I have guessed return value description; is it right?

File: libs/core/synchronization/include/hpx/synchronization/counting_semaphore.hpp (hunk 1 above)
Comment (hkaiser) [reply libs/core/synchronization/include/hpx/synchronization/counting_semaphore.hpp:364]: Yes, this looks right to me.

File: libs/core/synchronization/include/hpx/synchronization/counting_semaphore.hpp (hunk 2 above)
Comment (hkaiser) [reply libs/core/synchronization/include/hpx/synchronization/counting_semaphore.hpp:373]: Yes, all good. Thanks!

**Comment by hkaiser** [comment]
@bhumitattarde I have separated `hpx::binary_semaphore` into its own file. Also, please resolve the conflicts in this PR.

File: libs/core/synchronization/include/hpx/synchronization/counting_semaphore.hpp [hunk 3]
```diff
@@ -22,41 +22,379 @@
...
+    ///        \namedrequirement{MoveConstructible},
+    ///        \namedrequirement{CopyAssignable},
+    ///        or \namedrequirement{MoveAssignable}.
+    ///
+    /// \note \a counting_semaphore_var's \a try_acquire() can spuriously fail.
+    ///
+    /// \tparam Mutex Type of mutex
+    /// \tparam N     The initial value of the internal semaphore lock
+    ///               count.
+    ///
```
Comment (hkaiser) [inline libs/core/synchronization/include/hpx/synchronization/counting_semaphore.hpp:309]: `binary_semaphore` is now in its own file (as discussed). I think you should move the corresponding documentations there as well.

File: libs/core/synchronization/include/hpx/synchronization/counting_semaphore.hpp [hunk 4]
```diff
@@ -22,41 +22,379 @@
...
+
+        /// \copydoc hpx::counting_semaphore::acquire()
+        void acquire();
+
+        /// \copydoc hpx::counting_semaphore::try_acquire_until()
+        bool try_acquire_until(hpx::chrono::steady_time_point const& abs_time);
+
+        /// \copydoc hpx::counting_semaphore::try_acquire_for()
+        bool try_acquire_for(hpx::chrono::steady_duration const& rel_time);
+    };
```
Comment (hkaiser) [inline libs/core/synchronization/include/hpx/synchronization/counting_semaphore.hpp:249]: `binary_semaphore` is now in its own file (as discussed). I think you should move the corresponding documentations there as well.

**Comment by bhumitattarde** [comment]
@hkaiser should be okay now.

**Review by hkaiser — APPROVED** [review]
LGTM, thanks!

**Comment by hkaiser** [comment]
bors merge

**Pull request merged**